test/fuzz/fuzzerutil_test.cpp

```diff
@@ -0,0 +1,1519 @@
+         %34 = OpConstant %26 5
+         %35 = OpTypeArray %14 %34
+         %36 = OpTypePointer Function %35
+         %38 = OpConstant %14 5.5
+         %39 = OpConstant %14 4.4000001
+         %40 = OpConstant %14 3.29999995
+         %41 = OpConstant %14 2.20000005
+         %42 = OpConstant %14 1.10000002
+         %43 = OpConstantComposite %35 %38 %39 %40 %41 %42
+         %44 = OpConstant %26 3
+         %45 = OpTypeArray %20 %44
+         %46 = OpTypePointer Function %45
+         %48 = OpConstant %20 3
+         %49 = OpConstant %20 7
+         %50 = OpConstant %20 9
+         %51 = OpConstantComposite %45 %48 %49 %50
+         %53 = OpTypePointer Input %14
+         %54 = OpVariable %53 Input
+          %4 = OpFunction %2 None %3
+          %5 = OpLabel
+          %8 = OpVariable %7 Function
+         %10 = OpVariable %7 Function
+         %12 = OpVariable %7 Function
+         %13 = OpVariable %7 Function
+         %16 = OpVariable %15 Function
+         %18 = OpVariable %15 Function
+         %22 = OpVariable %21 Function
+         %24 = OpVariable %21 Function
+         %28 = OpVariable %27 Function
+         %30 = OpVariable %21 Function
+         %32 = OpVariable %21 Function
+         %37 = OpVariable %36 Function
+         %47 = OpVariable %46 Function
+               OpStore %8 %9
+               OpStore %10 %11
+               OpStore %12 %9
+               OpStore %13 %11
+               OpStore %16 %17
+               OpStore %18 %19
+               OpStore %22 %23
+               OpStore %24 %25
+               OpStore %28 %29
+               OpStore %30 %31
+               OpStore %32 %33
+               OpStore %37 %43
+               OpStore %47 %51
+               OpReturn
+               OpFunctionEnd
+  )";
+
+  const auto env = SPV_ENV_UNIVERSAL_1_4;
+  const auto consumer = nullptr;
+  const std::unique_ptr<opt::IRContext> context =
+      BuildModule(env, consumer, shader, kFuzzAssembleOption);
+  spvtools::ValidatorOptions validator_options;
+  ASSERT_TRUE(fuzzerutil::IsValidAndWellFormed(context.get(), validator_options,
+                                               kConsoleMessageConsumer));
+  TransformationContext transformation_context(
+      MakeUnique<FactManager>(context.get()), validator_options);
+
+  opt::IRContext* ir_context = context.get();
+
+  // The id of a boolean constant will be returned give boolean type id 6.
+  uint32_t maybe_bool_id = fuzzerutil::MaybeGetZeroConstant(
+      ir_context, transformation_context, 6, false);
+  // The id of a 32 bit float constant will be returned given the float type
+  // id 14.
+  uint32_t maybe_float_id = fuzzerutil::MaybeGetZeroConstant(
+      ir_context, transformation_context, 14, false);
+  uint32_t maybe_signed_int_id = fuzzerutil::MaybeGetZeroConstant(
+      ir_context, transformation_context, 20, false);
+  uint32_t maybe_unsigned_int_id = fuzzerutil::MaybeGetZeroConstant(
+      ir_context, transformation_context, 26, false);

```

Comment:
Note that there is no zero constant with type 26. The instruction with id 29 is a *constant* with type 26, but it is not a *zero* constant: it has the value 100.

Perhaps there are similar problems with the other assertions - I didn't check.

Please take a careful look and adapt accordingly.

Reply:
Thanks. I think I mistook the zero constant for the signed one with the unsigned one and had the error. Now its fixed.
However, here is the float part of the shader script: 
```
         %14 = OpTypeFloat 32
         %15 = OpTypePointer Function %14
         %17 = OpConstant %14 1.23000002
         %19 = OpConstant %14 1.11000001
         ... 
         %36 = OpConstant %26 5
         %37 = OpTypeArray %14 %36
         %38 = OpTypePointer Function %37
         %40 = OpConstant %14 5.5
         %41 = OpConstant %14 4.4000001
         %42 = OpConstant %14 3.29999995
         %43 = OpConstant %14 2.20000005
         %44 = OpConstant %14 1.10000002
```
In this case, ```  uint32_t maybe_float_id = fuzzerutil::MaybeGetZeroConstant(
      ir_context, transformation_context, 14, false);``` should have a value of either 17 or 19, since the type id of 14 represents a 32 bit float type, but in this case the assertion failed. Down below I also have an 5-element-array of floats. I first thought that it might return one of those ids (albeit unlikely),  but still it fails.
 I then ran 
```
  ASSERT_TRUE(std::find(float_ids.begin(), float_ids.end(), maybe_float_id)
    != float_ids.end()); 
```
and it turns out it cannot find a id with float type at all.

Reply:
In your above code, neither %17 nor %19 represent *zeros* of type float:

```
         %17 = OpConstant %14 1.23000002
```

This is the value 1.23000002 as a float.

```
         %19 = OpConstant %14 1.11000001
```

This is the value 1.11000001 as a float.

It is thus correct that `MaybeGetZeroConstant` doesn't return either of them.

A zero float constant would look like:

```
         %res = OpConstant %14 0
```

i.e., the value would be 0.

`MaybeGetZeroConstant` is implemented for floats. In the code you quoted:

```
    case SpvOpTypeFloat:
    case SpvOpTypeInt: {
      const auto width = type_inst->GetSingleWordInOperand(0);
      std::vector<uint32_t> words = {0};
      if (width > 32) {
        words.push_back(0);
      }

      return MaybeGetScalarConstant(ir_context, transformation_context, words,
                                    scalar_or_composite_type_id, is_irrelevant);
    }
```

the cases for `SpvOpTypeFloat` and `SpvOpTypeInt` lead to the same code being executed: if the opcode is `SpvOpTypeFloat` or `SpvOpTypeInt` then the code below these case lables is executed - that's because there is no `break` to stop code that takes the `SpvOpTypeFloat` case "falling through" to the `SpvOpTypeInt` case. So both cases are handled the same way.

Does that help clear things up?

Reply:
Got it. Resolved.